pkg/controller/clusterdeployment/clusterdeployment_controller.go

```diff
@@ -1246,6 +1278,52 @@ func (r *ReconcileClusterDeployment) deleteJobOnHashChange(existingJob, generate
 	return newJobNeeded, nil
 }
 
+// cleanupInstallLogPVC will immediately delete the PVC (should it exist) if the cluster was installed successfully, without retries.
+// If there were retries, it will delete the PVC if it has been more than 7 days since the job was completed.
+func (r *ReconcileClusterDeployment) cleanupInstallLogPVC(cd *hivev1.ClusterDeployment, existingJob *batchv1.Job, cdLog log.FieldLogger) error {
+	if !cd.Status.Installed {
+		return nil
+	}
+
+	if existingJob == nil {
+		return nil
+	}
+
+	pvc := &corev1.PersistentVolumeClaim{}
+	// PVC re-uses the name of the install job as they are related, and we know it should be unique.
+	err := r.Get(context.TODO(), types.NamespacedName{Name: existingJob.Name, Namespace: existingJob.Namespace}, pvc)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		cdLog.WithError(err).Error("error looking up install logs PVC")
+		return err
+	}
+
+	if cd.Status.InstallRestarts == 0 {
+		cdLog.WithField("pvc", pvc.Name).Info("deleting logs PersistentVolumeClaim for installed cluster with no restarts")
+		if err := r.Delete(context.TODO(), pvc); err != nil {
+			cdLog.WithError(err).Error("error deleting install logs PVC")
+			return err
+		}
+		return nil
+	}
+
+	// Otherwise, delete if more than 7 days have passed.
+	if time.Since(existingJob.Status.CompletionTime.Time) > (7 * 24 * time.Hour) {
+		cdLog.WithField("pvc", pvc.Name).Info("deleting logs PersistentVolumeClaim for cluster that was installed after restarts more than 7 days ago")
+		if err := r.Delete(context.TODO(), pvc); err != nil {
+			cdLog.WithError(err).Error("error deleting install logs PVC")
+			return err
+		}
+		return nil
+	}
+
+	cdLog.WithField("pvc", pvc.Name).Debug("preserving logs PersistentVolumeClaim for cluster with install restarts for 7 days")
```

Comment:
Are we relying on periodic resyncs of the clusterdeployment to pick up when the PVC has expired? Does kubebuilder do anything to swallow syncs when the revision of the clusterdeployment has not changed?

Reply:
We are relying on it yes, if kubebuilder is doing so then we need to rethink some things, I would be quite surprised if that were the case though. Reconciliation should be consistent and periodic even if nothing has changed in the desired state.